src/libraries/Common/src/Interop/OSX/Swift.Runtime/Foundation.cs

```diff
@@ -0,0 +1,217 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+#pragma warning disable CS3016 // Arrays as attribute arguments are not CLS Compliant
+#pragma warning disable SYSLIB1051
+
+namespace Swift.Runtime
+{
+    /// <summary>
+    /// Represents a Swift type in C#.
+    /// </summary>
+    internal unsafe interface ISwiftType
+    {
+        public abstract void* Metadata();
+    }
```

Comment:
Should this be `ISwiftObject` instead of `ISwiftType` as instances of this interface are Swift objects?

Alternatively, should Metadata be a static abstract (could it ever depend on the `this` instance or is it always a P/Invoke)?

A third option would be to have a `void* Payload` property and make the Metadata property static. Then we'd have the mechanisms to pass around unfrozen types generically.

Reply:
Considering that all types in Swift have type metadata associated with them and structs, classes, enums and actors have specific metadata accessor functions, it's best that this remains a more general name. I could see `ISwiftNominalType` as a more accurate name.

Reply:
I don't like "Type" being in the name where instances are actual values and not representations of types, unless the whole API is static abstracts (at which point it's an API of the type itself, not the instances of the type).